### PR TITLE
point search

### PR DESCRIPTION
**Pull request by pingxia0823**
https://app.clubhouse.io/radarlabs/story/1606/sdk-ios-point-search

### CODE REVIEW

File: RadarSDK/Radar.h [hunk 1]
```diff
@@ -121,56 +122,63 @@ typedef NS_ENUM(NSInteger, RadarRouteUnits) {
  
  @see https://radar.io/documentation/sdk#ios-foreground
  */
-typedef void(^ _Nullable RadarLocationCompletionHandler)(RadarStatus status, CLLocation * _Nullable location, BOOL stopped);
+typedef void (^_Nullable RadarLocationCompletionHandler)(RadarStatus status, CLLocation *_Nullable location, BOOL stopped);
```
Comment (pingxia0823) [inline RadarSDK/Radar.h]: I use a clang format CLI locally. 

However, it would format the entire file, not just the code blocks I added.

File: RadarSDK/Radar.h [hunk 2]
```diff
@@ -121,56 +122,63 @@ typedef NS_ENUM(NSInteger, RadarRouteUnits) {
...
 
  @see https://radar.io/documentation/api#search-geofences
  */
-typedef void(^ _Nonnull RadarSearchGeofencesCompletionHandler)(RadarStatus status, CLLocation * _Nullable location, NSArray<RadarGeofence *> * _Nullable geofences);
+typedef void (^_Nonnull RadarSearchGeofencesCompletionHandler)(RadarStatus status, CLLocation *_Nullable location, NSArray<RadarGeofence *> *_Nullable geofences);
+
+/**
+ Called when a point search request succeeds, fails, or times out. Receives the request status and, if successful, the location and an array of points sorted by distance.
+ 
+ // TODO link to docs here
```
Comment (pingxia0823) [inline RadarSDK/Radar.h]: TODO pending documentation

File: .gitignore [hunk 3]
```diff
@@ -7,3 +7,4 @@ sonar-reports/
 .scannerwork/
 DerivedData/
 build/
+.clang-format
```
Comment (pingxia0823) [inline .gitignore]: I use this clang-format

```
BasedOnStyle: Chromium

AlignTrailingComments: true
AllowShortIfStatementsOnASingleLine: false
BreakBeforeBraces: Linux
ColumnLimit: 0
IndentWidth: 4
KeepEmptyLinesAtTheStartOfBlocks: true
Language: ObjC
MaxEmptyLinesToKeep: 4
ObjCSpaceAfterProperty: true
ObjCSpaceBeforeProtocolList: false
PointerBindsToType: false
SpacesBeforeTrailingComments: 1
TabWidth: 4
UseTab: Never
```

File: RadarSDK.xcodeproj/xcshareddata/xcschemes/RadarSDKTests.xcscheme [hunk 4]
```diff
@@ -0,0 +1,52 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<Scheme
```
Comment (pingxia0823) [inline RadarSDK.xcodeproj/xcshareddata/xcschemes/RadarSDKTests.xcscheme]: Add a new schema for `RadarSKDTests` target, so we could do the testing within the xcode, with the benefit of individual case testing and debugging.

File: .gitignore (hunk 3 above)
Comment (jsani-radar) [reply .gitignore]: in your experience, what's the standard in iOS codebases / SDKs you've seen for standardizing code style and linting?

this seems pretty close to an `.eslintrc` that we could check in to keep our code standardized across the company and with open source contributions too

File: RadarSDK.xcodeproj/xcshareddata/xcschemes/RadarSDKTests.xcscheme (hunk 4 above)
Comment (jsani-radar) [reply RadarSDK.xcodeproj/xcshareddata/xcschemes/RadarSDKTests.xcscheme]: this is huge!!

File: RadarSDK/Radar.h [hunk 5]
```diff
@@ -121,56 +122,63 @@ typedef NS_ENUM(NSInteger, RadarRouteUnits) {
...
-typedef void(^ _Nonnull RadarSearchPlacesCompletionHandler)(RadarStatus status, CLLocation * _Nullable location, NSArray<RadarPlace *> * _Nullable places);
+typedef void (^_Nonnull RadarSearchPlacesCompletionHandler)(RadarStatus status, CLLocation *_Nullable location, NSArray<RadarPlace *> *_Nullable places);
 
 /**
  Called when a geofence search request succeeds, fails, or times out. Receives the request status and, if successful, the location and an array of geofences sorted by distance.
 
  @see https://radar.io/documentation/api#search-geofences
  */
-typedef void(^ _Nonnull RadarSearchGeofencesCompletionHandler)(RadarStatus status, CLLocation * _Nullable location, NSArray<RadarGeofence *> * _Nullable geofences);
+typedef void (^_Nonnull RadarSearchGeofencesCompletionHandler)(RadarStatus status, CLLocation *_Nullable location, NSArray<RadarGeofence *> *_Nullable geofences);
```
Comment (jsani-radar) [inline RadarSDK/Radar.h]: thinking through this a bit more, do you think it's possible to have a generic `RadarSearchCompletionHandler` that takes in an underlying entity class on which the operation searches? I can see that as a clean way to scale our search offering as we expand this out

File: RadarSDK/Radar.h [hunk 6]
```diff
@@ -431,17 +439,47 @@ typedef void(^ _Nonnull RadarRouteCompletionHandler)(RadarStatus status, RadarRo
 
  @param near The location to search.
  @param radius The radius to search, in meters. A number between 100 and 10000.
- @param tags An array of tags to filter. See https://radar.io/documentation/geofences
```
Comment (jsani-radar) [inline RadarSDK/Radar.h]: why was this removed? seems like the link would be helpful?

File: RadarSDK/Radar.h [hunk 7]
```diff
@@ -431,17 +439,47 @@ typedef void(^ _Nonnull RadarRouteCompletionHandler)(RadarStatus status, RadarRo
...
+                       tags:(NSArray *_Nullable)tags
                       limit:(int)limit
           completionHandler:(RadarSearchGeofencesCompletionHandler)completionHandler
     NS_SWIFT_NAME(searchGeofences(near:radius:tags:limit:completionHandler:));
 
+/**
+ Gets the device's current location, then searches for points near that location, sorted by distance.
+
+ @param radius The radius to search, in meters. A number between 100 and 10000.
+ @param tags An array of tags to filter.
```
Comment (jsani-radar) [inline RadarSDK/Radar.h]: we should probably have a placeholder here for the docs link as we do for geofences

File: RadarSDK/Radar.h [hunk 8]
```diff
@@ -431,17 +439,47 @@ typedef void(^ _Nonnull RadarRouteCompletionHandler)(RadarStatus status, RadarRo
...
+ @param limit The max number of points to return. A number between 1 and 100.
+ @param completionHandler A completion handler.
+*/
++ (void)searchPointsWithRadius:(int)radius
+                          tags:(NSArray *_Nullable)tags
```
Comment (jsani-radar) [inline RadarSDK/Radar.h]: I know this isn't the case on the other search endpoints, but any reason why we don't type this `NSArray`? we know it's gonna be `NSArray<NSString *>`, right?

File: RadarSDK/RadarAPIClient.m [hunk 9]
```diff
@@ -337,60 +367,111 @@ - (void)searchGeofencesNear:(CLLocation * _Nonnull)near
...
+                      NSArray<RadarGeofence *> *geofences = [RadarGeofence geofencesFromObject:geofencesObj];
+                      if (geofences) {
+                          return completionHandler(RadarStatusSuccess, res, geofences);
+                      }
+
+                      completionHandler(RadarStatusErrorServer, nil, nil);
+                    }];
+}
+
+- (void)searchPointsNear:(CLLocation *)near radius:(int)radius tags:(NSArray *)tags limit:(int)limit completionHandler:(RadarSearchPointsAPICompletionHandler)completionHandler
```
Comment (jsani-radar) [inline RadarSDK/RadarAPIClient.m]: tab down these params as we do with `searchGeofencesNear` above?

File: RadarSDKTests/Resources/search_points.json [hunk 10]
```diff
@@ -0,0 +1,62 @@
+{
+   "meta":{
+      "code":200,
+      "hasMore":false
```
Comment (jsani-radar) [inline RadarSDKTests/Resources/search_points.json]: interesting, this is the first time I've seen this meta key - is it for pagination?

File: RadarSDK/RadarPoint.m [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+
+@implementation RadarPoint
+
++ (NSArray<RadarPoint *> *)pointsFromObject:(id)object
+{
+    if (!object || ![object isKindOfClass:[NSArray class]]) {
+        return nil;
+    }
+
+    return [(NSArray *)object radar_mapObjectsUsingBlock:^RadarPoint *_Nullable(id _Nonnull pointObj) {
```
Comment (jsani-radar) [inline RadarSDK/RadarPoint.m]: I know this might be "cleaner" since it's using `map` and is therefore immutable, but the convention in our other models is to loop over the array -- https://github.com/radarlabs/radar-sdk-ios/blob/00b00af254bdee5d77d27b1ec7075809c0b3c25a/RadarSDK/RadarAddress.m#L21-L27

File: RadarSDK/RadarCollectionAdditions.m [hunk 12]
```diff
@@ -0,0 +1,66 @@
+//
```
Comment (jsani-radar) [inline RadarSDK/RadarCollectionAdditions.m]: oh these are so _sweet_ - can't wait to get these in all our models

File: .gitignore (hunk 3 above)
Comment (pingxia0823) [reply .gitignore]: yeah, the pretty standard practice, at least AFAIK, is have a .clang-format in the codebase, and everyone has that in git commit / push hook.

File: RadarSDK/Radar.h (hunk 5 above)
Comment (pingxia0823) [reply RadarSDK/Radar.h]: due to the restriction of ObjC, no generic is allowed for the typedef.... 

So we could either go verbose like this, or use a generic `NSArray *` for the callback. 

A slightly hacky way to avoid these boilerplate code is using macro

```
#define RADAR_SEARCH_COMPLETION_HANDLER(_contextType) ....

RADAR_SEARCH_COMPLETION_HANDLER(RadarGeofence);
RADAR_SEARCH_COMPLETION_HANDLER(RadarPoint)
```

File: RadarSDK/Radar.h (hunk 6 above)
Comment (pingxia0823) [reply RadarSDK/Radar.h]: no idea why, will revert.

File: RadarSDK/RadarAPIClient.m (hunk 9 above)
Comment (pingxia0823) [reply RadarSDK/RadarAPIClient.m]: must be some clang-format rule

File: RadarSDKTests/Resources/search_points.json (hunk 10 above)
Comment (pingxia0823) [reply RadarSDKTests/Resources/search_points.json]: Oh, sorry, search shall not return this field. I copied the JSON response from points get call (which has pagination)

**Comment by pingxia0823** [comment]
```
the review was slightly hard to isolate due to all the formatting changes, wondering if we should separate adding the points functionality with current state and then applying lint after into two PRs - just concerned I might've missed something
```

cool, Let me remove the auto formatting

File: RadarSDK/Radar.h [hunk 13]
```diff
@@ -151,6 +152,13 @@ typedef void(^ _Nonnull RadarSearchPlacesCompletionHandler)(RadarStatus status,
  */
 typedef void(^ _Nonnull RadarSearchGeofencesCompletionHandler)(RadarStatus status, CLLocation * _Nullable location, NSArray<RadarGeofence *> * _Nullable geofences);
 
+/**
+Called when a point search request succeeds, fails, or times out. Receives the request status and, if successful, the location and an array of points sorted by distance.
+
+@see https://radar.io/documentation/api#search-geofences
+*/
+typedef void (^_Nonnull RadarSearchPointsCompletionHandler)(RadarStatus status, CLLocation *_Nullable location, NSArray<RadarPoint *> *_Nullable points);
```
Comment (jsani-radar) [inline RadarSDK/Radar.h]: nit: other callbacks seem to have a space between `^` and `_` - this will likely get fixed / standardized with `.clang-format`

File: RadarSDK/RadarAPIClient.h [hunk 14]
```diff
@@ -27,6 +28,8 @@ typedef void(^ _Nullable RadarSearchPlacesAPICompletionHandler)(RadarStatus stat
 
 typedef void(^ _Nullable RadarSearchGeofencesAPICompletionHandler)(RadarStatus status, NSDictionary * _Nullable res, NSArray<RadarGeofence *> * _Nullable geofences);
 
+typedef void (^_Nullable RadarSearchPointsAPICompletionHandler)(RadarStatus status, NSDictionary *_Nullable res, NSArray<RadarPoint *> *_Nullable points);
```
Comment (jsani-radar) [inline RadarSDK/RadarAPIClient.h]: ditto

File: RadarSDK/RadarPoint.m (hunk 11 above)
Comment (jsani-radar) [reply RadarSDK/RadarPoint.m]: bumping this - do we want to move towards `mapObjects` over `for`?

File: RadarSDK/Radar.h (hunk 13 above)
Comment (pingxia0823) [reply RadarSDK/Radar.h]: yeah, this will get standardized in another PR

File: RadarSDK/RadarPoint.m (hunk 11 above)
Comment (pingxia0823) [reply RadarSDK/RadarPoint.m]: oops. let me update

File: RadarSDK/RadarCollectionAdditions.m [hunk 15]
```diff
@@ -0,0 +1,72 @@
...
+    if (!geoJsonDict) {
+        return nil;
+    }
+
+    NSArray *coordinates = [geoJsonDict radar_arrayForKey:@"coordinates"];
+    if (!coordinates || coordinates.count != 2 || ![coordinates[0] isKindOfClass:[NSNumber class]] || ![coordinates[1] isKindOfClass:[NSNumber class]]) {
+        return nil;
+    }
+    float longitude = [(NSNumber *)coordinates[0] floatValue];
+    if (longitude > 180 || longitude < -180) {
```
Comment (jsani-radar) [inline RadarSDK/RadarCollectionAdditions.m]: nice, I like these extra conditions
